capsules/src/alarm.rs

```diff
@@ -248,15 +248,30 @@ impl<'a, A: Alarm<'a>> SyscallDriver for AlarmDriver<'a, A> {
 
 impl<'a, A: Alarm<'a>> time::AlarmClient for AlarmDriver<'a, A> {
     fn alarm(&self) {
-        let now: Ticks32 = Ticks32::from(self.alarm.now().into_u32());
+        let now = self.alarm.now();
         self.app_alarms.each(|_processid, alarm, upcalls| {
             if let Expiration::Enabled { reference, dt } = alarm.expiration {
-                // Now is not within reference, reference + ticks; this timer
-                // as passed (since reference must be in the past)
-                if !now.within_range(
-                    Ticks32::from(reference),
-                    Ticks32::from(reference.wrapping_add(dt)),
-                ) {
+                // `now` is a 64-bit value, but we are only allowed a 32-bit
+                // `reference` or `dt`.
+                let within_range = if now < u32::MAX.into() {
+                    // The current time fits in a 32-bit value. We can compare
+                    // `now` with `reference + dt` where all values are 64-bit.
+                    now.within_range(
+                        A::Ticks::from(reference),
+                        A::Ticks::from((reference as u64).wrapping_add(dt as u64)),
+                    )
+                } else {
+                    // `now` has 'overflowed" the 32-bit value, so let's mask
+                    // off the top bits to represent a 32-bit overflow.
+                    let now = A::Ticks::from(now.into_u32());
+
+                    now.within_range(
+                        A::Ticks::from(reference),
+                        A::Ticks::from(reference).wrapping_add(dt.into()),
+                    )
+                };
```

Comment:
I'm confused why we need these two cases. If `now < u32::MAX.into()`, can't one still just execute line 266? In use, the variable `now` is always less than `u32::MAX` because its truncated in line 266. But then why does the top case use `u64` and the bottom cases use just `A::Ticks`?

Reply:
I haven't looked at this for awhile. I'm guessing that it's separated out to make the control flow clearer to understand what is going on and what we are comparing against.